### ISSUE TITLE
Feature request: mark functions `noexcept` in C++

### ISSUE THREAD

**Issue by rohanlean** [axboe/liburing#1117]
Hi,

first of all, many thanks for creating io_uring!

Because you seem to be using raw syscalls everywhere instead of POSIX wrappers that could be cancellation points, the functions in liburing thankfully do not throw and probably never will.  Yet they are not marked `noexcept`, which inhibits optimisations.

I propose that all functions be marked `noexcept`.

**Comment by isilence** [comment]
Apparently, `extern "C"` doesn't imply `noexcept` and it might be worth to add.
I wonder though, have you tried it? Any what's the difference in binary size / assembly?

**Comment by rohanlean** [comment]
Yes, it absolutely makes a difference.  A missing `noexcept` can prevent code motion, dead-code elimination, and a cascade of other optimisation.  For a simple example consider `T x; for (;;) io_uring_enter(...)`.  Unless the compiler knows that `io_uring_enter` does not throw, it needs to insert a call to (and possibly generate) the destructor of `T`, possibly insert a call to (and link) `std::terminate`, generate unwind tables, etc.

It is worth also marking the the functions with `__attribute__(nothrow)` in C (if supported), by the way.  Some C code is compiled with exceptions enabled, and it can make a difference there.  For example Fedora compiles everything with exceptions enabled.  You can look at how glibc defines its `__THROW` macro in `sys/cdefs.h`.

**Comment by rohanlean** [comment]
What are the plans for this?  It would be fairly easy to implement and has benefits, with the only downside being a slight increase of verbosity in the headers.  Here is an illustration of some of the effects that I alluded to in my last message: [example in the compiler explorer](https://godbolt.org/z/EfYzj59rh).

**Comment by DylanZA** [comment]
fwiw it looks like GLIBC does this for example on syscalls: 

https://github.com/bminor/glibc/blob/master/posix/unistd.h#L1091C7-L1091C8

defined here: https://github.com/bminor/glibc/blob/1944817240eeea5c044995e45a771a5a5193ecf1/misc/sys/cdefs.h#L86

I think its a good idea - perhaps @rohanlean can post a patch?

**Comment by ammarfaizi2** [comment]
That sounds good to me. It would be great if you could send a patch @rohanlean.
